software/OLED_BPM/OLED_BPM.ino

```diff
@@ -0,0 +1,240 @@
+uint8_t BPM = 0;
+bool IgnoreReading = false;
+bool FirstPulseDetected = false;
+unsigned long FirstPulseTime = 0;
+unsigned long SecondPulseTime = 0;
+unsigned long PulseInterval = 0;
+
+// Circular buffer to store pulse intervals
+CircularBuffer<int,30> buffer;
+
+void setup() {
+  // Initialize serial communication
+  Serial.begin(BAUD_RATE);
+  
+  // Set pin modes
+  pinMode(INPUT_PIN, INPUT);
+  pinMode(OUTPUT_PIN, OUTPUT);
+
+  // Initialize OLED display
+  if (!display.begin(SSD1306_SWITCHCAPVCC, OLED_Address)) {
+    Serial.println(F("SSD1306 allocation failed"));
+    for (;;);
+  }
+
+  // Set initial display settings
+  display.clearDisplay();
+  display.setTextSize(1);
+  display.setTextColor(WHITE);
+}
+
+void loop() {
+  // Calculate elapsed time since last loop
+  static unsigned long past = 0;
+  unsigned long present = micros();
+  unsigned long interval = present - past;
+  past = present;
+
+  // Timer for maintaining consistent sampling rate
+  static long timer = 0;
+  timer -= interval;
+
+  // Sample and process ECG signal at specified rate
+  if(timer < 0){
+    timer += 1000000 / SAMPLE_RATE;
+    
+    // Read sensor value and normalize
+    float sensor_value = analogRead(INPUT_PIN);
+    float signal = ECGFilter(sensor_value)/512;
+    
+    // Detect peaks in the signal
+    peak = Getpeak(signal);
+    
+    // Blink LED on detected peak
+    digitalWrite(OUTPUT_PIN, peak);
+
+    // Process detected peaks for BPM calculation
+    if(peak && IgnoreReading == false){
+        if(FirstPulseDetected == false){
+          FirstPulseTime = millis();
+          FirstPulseDetected = true;
+        }
+        else{
+          SecondPulseTime = millis();
+          PulseInterval = SecondPulseTime - FirstPulseTime;
+          buffer.unshift(PulseInterval);
+          FirstPulseTime = SecondPulseTime;
+        }
+        IgnoreReading = true;
+      }
+      if(!peak){
+        IgnoreReading = false;
+      }  
+      
+      // Calculate and display BPM when buffer is full
+      if (buffer.isFull()){
+        // Calculate average pulse interval
+        for(int i = 0 ;i < buffer.size(); i++){
+          avg+=buffer[i];
+        }
+        avg=avg/ buffer.size();
+        
+        // Calculate BPM
+        BPM = (1.0/avg) * 60.0 * 1000;
+        avg = 0;
+        buffer.pop();
+        
+        // Display BPM if it's within a reasonable range
+        if (BPM < 240){
+          Serial.print("BPM ");
+          Serial.println(BPM);
+          Serial.flush();
+          
+          // Update OLED display with new BPM
+          display.clearDisplay();
+          display.setTextSize(1);
+          display.setCursor(24, 3);
+          display.print("HEART_RATE(BPM): ");
+          display.setCursor(50,26);
+          display.setTextSize(3);
+          display.println(BPM);
+          display.display();
+        }
+      }  
+  }
+}
+
+// Function to detect peaks in the ECG signal
+bool Getpeak(float new_sample) {

```

Comment:
We need to provide a detailed explanation of this function.

Reply:
I have added few comments describing the detailed explanation of all attributes included in this function. Please review the same.